File: js/render.js
// Renders text to the sign's 1-bit-per-channel RGB bitfield format.
//
// The sign only supports 8 colors per pixel (R/G/B each on or off), packed
// 8 rows per byte, column-major, MSB = topmost row. This mirrors the
// reference driver's `get_separate_pixel_bytefields`.

function hexToRgb(hex) {
  const clean = hex.replace('#', '');
  const n = parseInt(clean, 16);
  return [(n >> 16) & 0xff, (n >> 8) & 0xff, n & 0xff];
}

// Text can embed per-segment colors as <#rrggbb>markers, e.g.
// "<#ff0000>HELLO<#00ff00>WORLD" renders HELLO in red and WORLD in green.
// A marker's color applies to everything after it until the next marker (or
// the end of the string) — there's no separate close tag.
const COLOR_START = '<';
const COLOR_END = '>';

function parseColorSegments(text) {
  const parts = [];
  for (const segment of text.split(COLOR_END)) {
    const pieces = segment.split(COLOR_START);
    if (pieces.length === 1) {
      parts.push({ color: null, text: pieces[0] });
    } else {
      parts.push({ color: pieces[pieces.length - 1], text: pieces.slice(0, -1).join('') });
    }
  }
  return parts;
}

/**
 * Draw `text` onto an offscreen canvas and pack it into RGB bitfields sized
 * to `outputHeight` rows (must be a multiple of 8). Returns the concatenated
 * R+G+B byte array plus the natural pixel width used.
 */
export function textToPixelBits(text, { color, backgroundColor, fontFamily, fontPx }, outputHeight) {
  if (outputHeight % 8 !== 0) {
    throw new Error('Sign height must be a multiple of 8');
  }

  const safeText = text.length > 0 ? text : ' ';
  const segments = parseColorSegments(safeText);

  const measure = document.createElement('canvas').getContext('2d');
  measure.font = `${fontPx}px ${fontFamily}`;
  let totalWidth = 0;
  for (const seg of segments) {
    if (seg.text.length > 0) totalWidth += measure.measureText(seg.text).width;
  }
  const textWidth = Math.max(1, Math.ceil(totalWidth));

  const padding = 1;
  const canvasWidth = textWidth + padding * 2;
  const canvasHeight = Math.max(outputHeight, Math.ceil(fontPx * 1.6));

  const canvas = document.createElement('canvas');
  canvas.width = canvasWidth;
  canvas.height = canvasHeight;
  const ctx = canvas.getContext('2d', { willReadFrequently: true });
  ctx.fillStyle = backgroundColor;
  ctx.fillRect(0, 0, canvasWidth, canvasHeight);
  ctx.font = `${fontPx}px ${fontFamily}`;
  ctx.textBaseline = 'top';

  let currentColor = color;
  let xOffset = padding;
  const y = Math.floor((canvasHeight - fontPx) / 2);
  for (const seg of segments) {
    if (seg.text.length > 0) {
      ctx.fillStyle = currentColor;
      ctx.fillText(seg.text, xOffset, y);
      xOffset += ctx.measureText(seg.text).width;
    }
    if (seg.color) {
      currentColor = seg.color;
    }
  }

  const imageData = ctx.getImageData(0, 0, canvasWidth, canvasHeight);
  const [bgR, bgG, bgB] = hexToRgb(backgroundColor);
  const topOffset = Math.floor((outputHeight - canvasHeight) / 2);

  const barrR = [];
  const barrG = [];
  const barrB = [];
  let tmpR = 0;
  let tmpG = 0;
  let tmpB = 0;

  for (let x = 0; x < canvasWidth; x++) {
    for (let y = 0; y < outputHeight; y++) {
      const srcY = y - topOffset;
      let r;
      let g;
      let b;
      if (srcY < 0 || srcY >= canvasHeight) {
        r = bgR;
        g = bgG;
        b = bgB;
      } else {
        const idx = (srcY * canvasWidth + x) * 4;
        r = imageData.data[idx];
        g = imageData.data[idx + 1];
        b = imageData.data[idx + 2];
      }

      tmpR = (tmpR << 1) | Math.round(r / 255);
      tmpG = (tmpG << 1) | Math.round(g / 255);
      tmpB = (tmpB << 1) | Math.round(b / 255);

      if (y % 8 === 7) {
        barrR.push(tmpR & 0xff);
        barrG.push(tmpG & 0xff);
        barrB.push(tmpB & 0xff);
        tmpR = 0;
        tmpG = 0;
        tmpB = 0;
      }
    }
  }

  return {
    pixelBits: new Uint8Array([...barrR, ...barrG, ...barrB]),
    width: canvasWidth,
    height: outputHeight,
  };
}

/** Draw a live preview of what the sign will show, upscaled for visibility. */
export function drawPreview(previewCanvas, text, options, deviceWidth, deviceHeight) {
  const { pixelBits, width, height } = textToPixelBits(text, options, deviceHeight);
  const planeSize = (width * height) / 8;
  const rPlane = pixelBits.slice(0, planeSize);
  const gPlane = pixelBits.slice(planeSize, planeSize * 2);
  const bPlane = pixelBits.slice(planeSize * 2, planeSize * 3);

  const scale = Math.max(1, Math.floor(600 / Math.max(width, deviceWidth || width)));
  previewCanvas.width = width * scale;
  previewCanvas.height = height * scale;
  const ctx = previewCanvas.getContext('2d');
  ctx.imageSmoothingEnabled = false;
  ctx.fillStyle = '#000';
  ctx.fillRect(0, 0, previewCanvas.width, previewCanvas.height);

  for (let x = 0; x < width; x++) {
    for (let y = 0; y < height; y++) {
      const byteIdx = x * (height / 8) + Math.floor(y / 8);
      const bitIdx = 7 - (y % 8);
      const r = (rPlane[byteIdx] >> bitIdx) & 1;
      const g = (gPlane[byteIdx] >> bitIdx) & 1;
      const b = (bPlane[byteIdx] >> bitIdx) & 1;
      if (r || g || b) {
        ctx.fillStyle = `rgb(${r * 255}, ${g * 255}, ${b * 255})`;
        ctx.fillRect(x * scale, y * scale, scale, scale);
      }
    }
  }

  return { width, height };
}

File: desktop/led_bridge/render.py
"""Text-to-pixel-bits rendering for the CoolLED1248 / CoolLEDX signs.

Ported from js/render.js. The sign only supports 8 colors per pixel (R/G/B
each on or off), packed 8 rows per byte, column-major, MSB = topmost row.

Browser <canvas> text rendering can't be reproduced pixel-for-pixel outside
a browser, so this uses Pillow with a bundled DejaVu Sans font (supports
Greek + the euro sign) as a close, dependency-free approximation.
"""

from __future__ import annotations

import base64
import io
import math
from functools import lru_cache

from PIL import Image, ImageDraw, ImageFont

from ._paths import resource_path

_FONT_PATH = resource_path("fonts", "DejaVuSans.ttf")

COLOR_START = "<"
COLOR_END = ">"


def hex_to_rgb(hex_color: str) -> tuple[int, int, int]:
    clean = hex_color.lstrip("#")
    n = int(clean, 16)
    return ((n >> 16) & 0xFF, (n >> 8) & 0xFF, n & 0xFF)


def parse_color_segments(text: str) -> list[dict]:
    """Split text on <#rrggbb> markers; a marker's color applies to
    everything after it until the next marker (or the end of the string).
    """
    parts = []
    for segment in text.split(COLOR_END):
        pieces = segment.split(COLOR_START)
        if len(pieces) == 1:
            parts.append({"color": None, "text": pieces[0]})
        else:
            parts.append({"color": pieces[-1], "text": "".join(pieces[:-1])})
    return parts


@lru_cache(maxsize=32)
def _font(font_px: int) -> ImageFont.FreeTypeFont:
    return ImageFont.truetype(_FONT_PATH, font_px)


def text_to_pixel_bits(
    text: str,
    color: str,
    background_color: str,
    font_px: int,
    output_height: int,
) -> tuple[bytes, int, int]:
    """Render `text` and pack it into RGB bitfields sized to `output_height`
    rows (must be a multiple of 8). Returns (pixel_bits, width, height).
    """
    if output_height % 8 != 0:
        raise ValueError("Sign height must be a multiple of 8")

    safe_text = text if len(text) > 0 else " "
    segments = parse_color_segments(safe_text)

    font = _font(font_px)
    total_width = 0.0
    for seg in segments:
        if seg["text"]:
            total_width += font.getlength(seg["text"])
    text_width = max(1, math.ceil(total_width))

    padding = 1
    canvas_width = text_width + padding * 2
    canvas_height = max(output_height, math.ceil(font_px * 1.6))

    image = Image.new("RGB", (canvas_width, canvas_height), hex_to_rgb(background_color))
    draw = ImageDraw.Draw(image)

    current_color = color
    x_offset = float(padding)
    y = (canvas_height - font_px) // 2
    for seg in segments:
        if seg["text"]:
            draw.text((x_offset, y), seg["text"], font=font, fill=hex_to_rgb(current_color), anchor="la")
            x_offset += font.getlength(seg["text"])
        if seg["color"]:
            current_color = seg["color"]

    pixels = image.load()
    bg_r, bg_g, bg_b = hex_to_rgb(background_color)
    top_offset = (output_height - canvas_height) // 2

    barr_r = bytearray()
    barr_g = bytearray()
    barr_b = bytearray()

    for x in range(canvas_width):
        tmp_r = tmp_g = tmp_b = 0
        for y in range(output_height):
            src_y = y - top_offset
            if src_y < 0 or src_y >= canvas_height:
                r, g, b = bg_r, bg_g, bg_b
            else:
                r, g, b = pixels[x, src_y]

            tmp_r = ((tmp_r << 1) | (1 if r >= 128 else 0)) & 0xFF
            tmp_g = ((tmp_g << 1) | (1 if g >= 128 else 0)) & 0xFF
            tmp_b = ((tmp_b << 1) | (1 if b >= 128 else 0)) & 0xFF

            if y % 8 == 7:
                barr_r.append(tmp_r)
                barr_g.append(tmp_g)
                barr_b.append(tmp_b)
                tmp_r = tmp_g = tmp_b = 0

    pixel_bits = bytes(barr_r) + bytes(barr_g) + bytes(barr_b)
    return pixel_bits, canvas_width, output_height


def render_preview_png_data_uri(
    text: str,
    color: str,
    background_color: str,
    font_px: int,
    device_width: int,
    device_height: int,
    max_pixels: int = 600,
) -> str:
    """Render exactly what would be sent to the sign, upscaled for
    visibility, as a data: URI so the frontend can show a byte-accurate
    preview without duplicating any rendering/parsing logic in JS.
    """
    pixel_bits, width, height = text_to_pixel_bits(text, color, background_color, font_px, device_height)
    plane_size = (width * height) // 8
    r_plane = pixel_bits[0:plane_size]
    g_plane = pixel_bits[plane_size : plane_size * 2]
    b_plane = pixel_bits[plane_size * 2 : plane_size * 3]

    image = Image.new("RGB", (width, height), (0, 0, 0))
    pixels = image.load()
    rows_per_byte = height // 8
    for x in range(width):
        for y in range(height):
            byte_idx = x * rows_per_byte + (y // 8)
            bit_idx = 7 - (y % 8)
            r = (r_plane[byte_idx] >> bit_idx) & 1
            g = (g_plane[byte_idx] >> bit_idx) & 1
            b = (b_plane[byte_idx] >> bit_idx) & 1
            pixels[x, y] = (r * 255, g * 255, b * 255)

    scale = max(1, max_pixels // max(width, device_width or width))
    image = image.resize((width * scale, height * scale), Image.NEAREST)

    buf = io.BytesIO()
    image.save(buf, format="PNG")
    encoded = base64.b64encode(buf.getvalue()).decode("ascii")
    return f"data:image/png;base64,{encoded}"
